rest/multipart.go

```diff
@@ -27,30 +27,56 @@ const kMaxInlineAttachmentSize = 200
 // JSON bodies smaller than this won't be GZip-encoded.
 const kMinCompressedJSONSize = 300
 
-// Parses a JSON MIME body, unmarshaling it into "into".
+// ReadJSONFromMIME parses a JSON MIME body, unmarshalling it into "into".
 // Closes the input io.ReadCloser once done.
 func ReadJSONFromMIME(headers http.Header, input io.ReadCloser, into interface{}) error {
-	contentType := headers.Get("Content-Type")
-	if contentType != "" && !strings.HasPrefix(contentType, "application/json") {
-		return base.HTTPErrorf(http.StatusUnsupportedMediaType, "Invalid content type %s", contentType)
+	// Performs the Content-Type validation and Content-Encoding check.
+	input, err := processContentEncoding(headers, input)
+	if err != nil {
+		return err
 	}
 
-	switch headers.Get("Content-Encoding") {
-	case "gzip":
-		var err error
-		if input, err = gzip.NewReader(input); err != nil {
-			return err
+	// Decode the body bytes into target structure.
+	err = base.JSONDecoderWithOptions(input, true, true).Decode(&into)
```

Comment:
I think this code is more opaque than the previous four lines of code it's replacing.  With the previous case, it was clear at a glance what options were being used for the decoder - the updated version relies on remembering the parameter order and whether true means 'allow unknown' or 'disallow unknown'.

Reply:
It did strike my mind in the first place that this code is less readable than the older version. But later I thought this would be reused across both the places while addressing the previous comments from @bbrks. Looking at this further, I think we can keep the old approach and changes are included in the latest commit.